### PR TITLE
refactor: rewrite using hooks

### LINKED ISSUE #1: Rewrite using React Hooks
Currently uses class syntax, but for that kind of feature Hooks look better choice.

### PR DESCRIPTION
**Pull request by Yaxian**

### CODE REVIEW

**Comment by moubi** [comment]
@Yaxian thanks for your help on that big step towards `v2` 🔥

There are some conflicts probably because of me ended up with some changes on master. Will probably be able to take a deeper look at your PR some time today or tomorrow.

**Comment by Yaxian** [comment]
OK, I will fix the conflicts.

File: index.d.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
+import * as React from "react";
```
Comment (moubi) [inline index.d.ts:1]: I have started reviewing the PR. Will create a test case with codesandbox as well.

First thing to ask in advance: isn't the presence of `index.d.ts` in place of `PropTypes` assumes we need to be in TypeScript environment?

Shouldn't we use both in order to cover both javascript and TypeScript environments?

**Comment by moubi** [comment]
@Yaxian can you fork this [codesandbox](https://codesandbox.io/s/swiping-views-u49lp) and use your `swipeable-react` fork to create a working example for the hooks?

**Comment by moubi** [comment]
I managed to create a codesandbox with your hook version [here](https://codesandbox.io/s/testing-version-with-hooks-ipt2k). The component is under `src/Swipeable.js`.

You can use it if it make sense.

File: src/Swipeable.js [hunk 2]
```diff
@@ -93,22 +75,34 @@ export default class Swipeable extends PureComponent {
...
+    el.addEventListener("touchend", handleTouchEnd);
+
+    return () => {
+      el.removeEventListener("touchstart", handleTouchStart);
+      el.removeEventListener("touchmove", handleTouchMove);
+      el.removeEventListener("touchend", handleTouchEnd);
+      instanceRef.current = null;
+      if (ref) ref.current = null;
+    };
+  }, []);
```
Comment (moubi) [inline src/Swipeable.js]: Isn't React going to complain about `ref` and `handleTouchEnd` missing in the dependency array?

Comment (Yaxian) [reply src/Swipeable.js]: I fixed the dependency warning. I did not use the `eslint` before.

File: src/Swipeable.js [hunk 3]
```diff
@@ -1,106 +1,123 @@
-// Based on https://github.com/john-doherty/swiped-events
-import { PureComponent } from "react";
+import { useEffect, useRef, forwardRef, useCallback } from "react";
 import PropTypes from "prop-types";
 
-export default class Swipeable extends PureComponent {
-  constructor(props) {
-    super(props);
+const Swipeable = forwardRef(function Swipeable(props, ref) {
```
Comment (moubi) [inline src/Swipeable.js:4]: @Yaxian looks better now.

What would happen if we completely remove the `forwardRef` and `ref` usage? 

So exporting something like:
```javascript
function Swipeable(props) {
  ...
}
```

I have update this [codesandbox](https://codesandbox.io/s/testing-version-with-hooks-ipt2k?file=/src/Swipeable.js) with your latest code.

Comment (Yaxian) [reply src/Swipeable.js:4]: Tests will fail and `function component` can not pass `ref` props.

Comment (moubi) [reply src/Swipeable.js:4]: Ok, does that mean any consumer component will break? 
Using it like the way it is now
```javascript
<Swipeable
        onSwipeLeft={this.handleSwipeLeft}
        onSwipeRight={this.handleSwipeRight}
      >
        {innerRef => (
          <div ref={innerRef}>
            ...
          </div>
        )}
      </Swipeable>
```
isn't `innerRef` going to be passed correctly without `forwardRef ` usage?

I briefly tested in the above codesandbox without `ref` and it seems to work fine. Can you confirm that?

Comment (Yaxian) [reply src/Swipeable.js:4]: We can not pass `ref` directly to `Swipeable`:
```
<Swipeable ref={ref} onSwipeLeft={() => {}}>
   {innerRef => (
          <div ref={innerRef}>
            ...
          </div>
        )}
</Swipeable>
```
if we do not use `forwardRef` to wrap a `function component`.

`innerRef ` still works fine.

Comment (moubi) [reply src/Swipeable.js:4]: Great, I think it's not necessary to change the user facing API of Swipeable for this rewrite, so let's keep the usage of `innerRef` and skip `forwardRef`.

Comment (Yaxian) [reply src/Swipeable.js:4]: the tests will fail  if we do not use forwardRef, cause we can not get the instance and its methods of `function component` like `class component`

Comment (moubi) [reply src/Swipeable.js:4]: That's true. Let's keep the `forwardRef ` then. 

I'll merge the MR soon.

**Comment by moubi** [comment]
Thank you again for your contribution @Yaxian. Merging...

**Pull request merged**